localstack/services/s3/provider.py

```diff
@@ -419,6 +669,285 @@ def is_object_expired(context: RequestContext, bucket: BucketName, key: ObjectKe
+    entry = {
+        "Source": "aws.s3",
+        "Resources": [f"arn:aws:s3:::{bucket.name}"],
+        "Detail": {
+            "version": "0",
+            "bucket": {"name": bucket.name},
+            "object": {
+                "key": key.name,
+                "size": key.size,
+                "etag": key.etag.strip('"'),
+                "sequencer": "0062E99A88DC407460",
+            },
+            "request-id": "RKREYG1RN2X92YX6",
+            "requester": "074255357339",
+            "source-ip-address": "127.0.0.1",  # TODO previously headers.get("X-Forwarded-For", "127.0.0.1").split(",")[0]
+        },
+    }
+    # messages are bit different for EventBridge, see https://docs.aws.amazon.com/AmazonS3/latest/userguide/EventBridge.html
+    if "ObjectCreated" in event_name:
+        entry["DetailType"] = "Object Created"
+        event_type = event_name[event_name.rindex(":") + 1 :]
+        if event_type in ["Put", "Post", "Copy"]:
+            event_type = f"{event_type}Object"
+        entry["Detail"]["reason"] = event_type
+
+    if "ObjectRemoved" in event_name:
+        entry["DetailType"] = "Object Deleted"
+        entry["Detail"]["reason"] = "DeleteObject"
+        entry["Detail"]["deletion-type"] = "Permanently Deleted"
+        entry["Detail"]["object"].pop("etag")
+        entry["Detail"]["object"].pop("size")
+
+    if "ObjectTagging" in event_name:
+        entry["DetailType"] = "Object Tags Added" if "Put" in event_name else "Object Tags Deleted"
+
+    entry["Detail"] = json.dumps(entry["Detail"])
+
+    try:
+        events_client.put_events(Entries=[entry])
+    except Exception as e:
+        LOG.exception(f'Unable to send notification for S3 bucket "{bucket}" to EventBridge', e)
+
+
+def _send_event_message(
+    event_name: str,
+    bucket: moto_s3_models.FakeBucket,
+    key: moto_s3_models.FakeKey,
+    notifications: NotificationConfiguration,
+    xray: str = None,
+):
+    for notification in notifications.get("QueueConfigurations", {}):
+        if _matching_event(notification["Events"], event_name) and _matching_filter(
+            notification.get("Filter"), key.name
+        ):
+            _send_event_message_to_queue(notification, event_name, bucket, key, xray)
+
+    for notification in notifications.get("TopicConfigurations", {}):
+        if _matching_event(notification["Events"], event_name) and _matching_filter(
+            notification.get("Filter"), key.name
+        ):
+            _send_event_message_to_topic(notification, event_name, bucket, key, xray)
+
+    for notification in notifications.get("LambdaFunctionConfigurations", {}):
+        if _matching_event(notification["Events"], event_name) and _matching_filter(
+            notification.get("Filter"), key.name
+        ):
+            _send_event_message_to_lambda(notification, event_name, bucket, key, xray)
+
+    if "EventBridgeConfiguration" in notifications:
+        _send_event_to_event_bridge(event_name, bucket, key)
```

Comment:
here i think the `Notifier` abstraction will come in handy, since there is no need to dispatch manually anymore.
conceptually the method will look something like this:

```python
notifiers = {
    "TopicConfigurations:" SnsNotifier(),
    "LambdaFunctionConfigurations": LambdaNotifier(),
    # ...
}

for type, notification_configs in notifications.items():
    for config in notification_configs:
        notifiers[type].notify(ctx, config) # <- this call should happen asynchronously, so we should add it to an executor queue, for example by using a ThreadPoolExecutor
```

Reply:
Do you think we should create a `ThreadPoolExecutor` for each type of notifications?
I would like to have some way to batch all notification events, then dispatch them into one call, to avoid creating multiple executors. One S3 operation = one executor for now?
It would mean that all `Notifiers` would kind of queue their work in a list, and after all events are created, we should dispatch them?
So one main `NotificationDispatcher`, containing the different type of `Notifiers`. What do you think?

Reply:
i conceptually it should be:

```python
notifiers = {
    "TopicConfigurations:" SnsNotifier(),
    "LambdaFunctionConfigurations": LambdaNotifier(),
    # ...
}

executor = ThreadPoolExecutor(num_threads)

def do_notify(notifiers, notifications, ctx):
    for type, notification_configs in notifications.items():
        for config in notification_configs:
            executor.submit(notifiers[type].notify, args=(ctx, config))
```

where num_threads could be a low number for now
this way, the call to do_notify is non-blocking and can be called from within the provider.